src/TodoContext.tsx

```diff
@@ -0,0 +1,166 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import React, { useReducer, useState, useEffect } from 'react';
+import { Filter, TodoType } from './types/types';
+
+const ADD_TODO = 'add todo';
+const REMOVE_TODO = 'remove todo';
+const COMPLETE_TODO = 'complete todo';
+const COMPLETE_ALL_TODO = 'complete all todo';
+const UPDATE_TODO = 'update todo';
+
+type Action = { type: typeof ADD_TODO, payLoad: TodoType }
+| { type: typeof REMOVE_TODO, payLoad: number | 'All' }
+| { type: typeof COMPLETE_TODO, payLoad: number }
+| { type: typeof COMPLETE_ALL_TODO }
+| { type: typeof UPDATE_TODO, payLoad: TodoType };
+
+function reducer(state: TodoType[], action: Action): TodoType[] {
+  switch (action.type) {
+    case ADD_TODO:
+      return [...state, action.payLoad];
+
+    case REMOVE_TODO: {
+      if (action.payLoad === 'All') {
+        const clearAllCompleted = state
+          .filter((todo: TodoType) => todo.completed !== true);
+
+        return clearAllCompleted;
+      }
+
+      const availableTodos = state
+        .filter((todo: TodoType) => todo.id !== action.payLoad);
+
+      return availableTodos;
+    }
+
+    case COMPLETE_TODO: {
+      const toggleComplete = state
+        .map((todo: TodoType) => {
+          if (todo.id === action.payLoad) {
+            return { ...todo, completed: !todo.completed };
+          }
+
+          return todo;
+        });
+
+      return toggleComplete;
+    }
+
+    case UPDATE_TODO: {
+      const updatedTodos = state.map(todo => {
+        return { ...todo };
+      });
+      const index = updatedTodos
+        .findIndex(todo => todo.id === action.payLoad.id);
+
+      updatedTodos.splice(index, 1, action.payLoad);
+
+      return updatedTodos;
+    }
+
+    case COMPLETE_ALL_TODO: {
+      const isCompleted = state.some(todo => todo.completed === true);
+      const allCompleted = state.every(todo => todo.completed === true);
```

Comment:
```suggestion
      const isCompleted = state.some(({ completed })=> completed);
      const allCompleted = state.every(({ completed })=> completed);
```

Reply:
Your comments noted. Thank you.
All done.